biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotations.java

```diff
@@ -26,55 +50,150 @@ public boolean analyzeJar(Analyzer analyzer) throws Exception {
+		if (v != null) {
+			if (definition.version.compareTo(v) > 0) {
+				SetLocation error = analyzer
+						.error("Generating XML for %s in type %s that uses a namespace version %s while you are building against %s",
+								definition.name, definition.implementation, definition.version, v);
+
+				error.details(analyzer.getPackageRef("org/osgi/service/component"));
+				analyzer.setTypeLocation(error, definition.implementation);
+			}
+		}
+	}
+
+	private Version getPackageVersion(Analyzer analyzer) {
+		PackageRef component = analyzer.getPackageRef("org/osgi/service/component");
 		Attrs attrs = analyzer.getClasspathExports().get(component);
 		if (attrs != null) {
 			String version = attrs.getVersion();
 			if (version != null && Verifier.isVersion(version)) {
-				Version v = new Version(version);
-				if (definition.version.compareTo(v) > 0) {
-					SetLocation error = analyzer
-							.error("Generating XML for %s in type %s that uses a namespace version %s while you are building against %s",
-									definition.name, definition.implementation, definition.version, v);
-
-					error.details(component);
-					analyzer.setTypeLocation(error, definition.implementation);
-				}
+				return new Version(version);
+			}
+		}
+		return null;
+	}
+
+	private void addProvidesHeader(String[] objectClass, Set<String> provides) {
+		if (objectClass.length > 0) {
+			Parameters p = new Parameters();
+			Attrs a = new Attrs();
+			StringBuilder sb = new StringBuilder();
+			String sep = "";
+			for (String oc : objectClass) {
+				sb.append(sep).append(oc);
+				sep = ",";
 			}
+			a.put("objectClass:List<String>", sb.toString());
+			p.put("osgi.service", a);
+			String s = p.toString();
+			provides.add(s);
 		}
 	}
 
+	private void addRequireHeader(String objectClass, Set<String> requires) {
+		Parameters p = new Parameters();
+		Attrs a = new Attrs();
+		a.put("filter:", "\"(objectClass=" + objectClass + ")\"");
+		a.put("effective:", "\"active\"");
+		p.put("osgi.service", a);
+		String s = p.toString();
+		requires.add(s);
+	}
+
+	private void addExtenRequireHeader(Set<String> requires) {
+		Parameters p = new Parameters();
+		Attrs a = new Attrs();
+		a.put("filter:", "\"(&(osgi.extender=osgi.component)(version>=1.3)(!(version>=2.0)))\"");
+		p.put("osgi.extender", a);
+		String s = p.toString();
+		requires.add(s);
+	}
+
+	/*
+	 * This method is a pass thru for the properties of the analyzer. If we have
+	 * such a header, we get the analyzer header and concatenate our values
+	 * after removing dups.
+	 */
+
+	public void updateHeader(Analyzer analyzer, String name, Set<String> set) {
+		String value = analyzer.getProperty(name);
+		if (!set.isEmpty()) {
+			//
+			// Remove duplicates and sort
+			//
+			String header = Strings.join(set);
+			if (value == null)
+				analyzer.setProperty(name, header);
+			else
+				analyzer.setProperty(name, value + "," + header);
```

Comment:
What about duplication coalescing? The user could have set service requirement or capabilities in the bnd file.